### ISSUE TITLE
exporter/stats/prometheus: java.lang.NoSuchFieldError: timestampMs

### ISSUE THREAD

**Issue by odeke-em** [census-instrumentation/opencensus-java#1215]
While running a demo for "OpenCensus for Java developers" https://github.com/orijtech/opencensus-for-grpc-java-developers, on enabling Prometheus
by
```diff
diff --git a/src/main/java/io/ocgrpc/capitalize/CapitalizeServer.java b/src/main/java/io/ocgrpc/capitalize/CapitalizeServer.java
index 9f2f851..f0ff7ac 100644
--- a/src/main/java/io/ocgrpc/capitalize/CapitalizeServer.java
+++ b/src/main/java/io/ocgrpc/capitalize/CapitalizeServer.java
@@ -142,7 +142,7 @@ public class CapitalizeServer {
             .setProjectId(gcpProjectId)
             .build());
 
-        if (false) {
+        if (true) {
             // And then the Prometheus exporter too
             PrometheusStatsCollector.createAndRegister();
             // Start the Prometheus server
```

and then running it by
```shell
$ GOOGLE_APPLICATION_CREDENTIALS=~/Downloads/census-demos-237a8e1e41df.json mvn exec:java -Dexec.mainClass=io.ocgrpc.capitalize.CapitalizeServer
```

and running prometheus with this config file
```shell
global:
  scrape_interval: 10s

  external_labels:
    monitor: 'ocgrpc-java' 

scrape_configs:
  - job_name: 'ocgrpc-java'

    scrape_interval: 10s

    static_configs:
      - targets: ['localhost:9821']
```
```shell
$ prometheus --config.file=prometheus.yml
```

I get a bunch of error logs and seeming crashes
```shell
[WARNING] 
java.lang.NoSuchFieldError: timestampMs
    at io.prometheus.client.exporter.common.TextFormat.write004 (TextFormat.java:49)
    at io.prometheus.client.exporter.HTTPServer$HTTPMetricHandler.handle (HTTPServer.java:59)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:77)
    at sun.net.httpserver.AuthFilter.doFilter (AuthFilter.java:82)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:80)
    at sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle (ServerImpl.java:685)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:77)
    at sun.net.httpserver.ServerImpl$Exchange.run (ServerImpl.java:657)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1167)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:641)
    at java.lang.Thread.run (Thread.java:844)
[WARNING] 
java.lang.NoSuchFieldError: timestampMs
    at io.prometheus.client.exporter.common.TextFormat.write004 (TextFormat.java:49)
    at io.prometheus.client.exporter.HTTPServer$HTTPMetricHandler.handle (HTTPServer.java:59)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:77)
    at sun.net.httpserver.AuthFilter.doFilter (AuthFilter.java:82)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:80)
    at sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle (ServerImpl.java:685)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:77)
    at sun.net.httpserver.ServerImpl$Exchange.run (ServerImpl.java:657)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1167)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:641)
    at java.lang.Thread.run (Thread.java:844)
[WARNING] 
java.lang.NoSuchFieldError: timestampMs
    at io.prometheus.client.exporter.common.TextFormat.write004 (TextFormat.java:49)
    at io.prometheus.client.exporter.HTTPServer$HTTPMetricHandler.handle (HTTPServer.java:59)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:77)
    at sun.net.httpserver.AuthFilter.doFilter (AuthFilter.java:82)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:80)
    at sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle (ServerImpl.java:685)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:77)
    at sun.net.httpserver.ServerImpl$Exchange.run (ServerImpl.java:657)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1167)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:641)
    at java.lang.Thread.run (Thread.java:844)
[WARNING] 
java.lang.NoSuchFieldError: timestampMs
    at io.prometheus.client.exporter.common.TextFormat.write004 (TextFormat.java:49)
    at io.prometheus.client.exporter.HTTPServer$HTTPMetricHandler.handle (HTTPServer.java:59)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:77)
    at sun.net.httpserver.AuthFilter.doFilter (AuthFilter.java:82)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:80)
    at sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle (ServerImpl.java:685)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:77)
    at sun.net.httpserver.ServerImpl$Exchange.run (ServerImpl.java:657)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1167)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:641)
    at java.lang.Thread.run (Thread.java:844)
[WARNING] 
java.lang.NoSuchFieldError: timestampMs
    at io.prometheus.client.exporter.common.TextFormat.write004 (TextFormat.java:49)
    at io.prometheus.client.exporter.HTTPServer$HTTPMetricHandler.handle (HTTPServer.java:59)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:77)
    at sun.net.httpserver.AuthFilter.doFilter (AuthFilter.java:82)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:80)
    at sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle (ServerImpl.java:685)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:77)
    at sun.net.httpserver.ServerImpl$Exchange.run (ServerImpl.java:657)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1167)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:641)
    at java.lang.Thread.run (Thread.java:844)
[WARNING] 
java.lang.NoSuchFieldError: timestampMs
    at io.prometheus.client.exporter.common.TextFormat.write004 (TextFormat.java:49)
    at io.prometheus.client.exporter.HTTPServer$HTTPMetricHandler.handle (HTTPServer.java:59)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:77)
    at sun.net.httpserver.AuthFilter.doFilter (AuthFilter.java:82)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:80)
    at sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle (ServerImpl.java:685)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:77)
    at sun.net.httpserver.ServerImpl$Exchange.run (ServerImpl.java:657)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1167)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:641)
    at java.lang.Thread.run (Thread.java:844)
[WARNING] 
java.lang.NoSuchFieldError: timestampMs
    at io.prometheus.client.exporter.common.TextFormat.write004 (TextFormat.java:49)
    at io.prometheus.client.exporter.HTTPServer$HTTPMetricHandler.handle (HTTPServer.java:59)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:77)
    at sun.net.httpserver.AuthFilter.doFilter (AuthFilter.java:82)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:80)
    at sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle (ServerImpl.java:685)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:77)
    at sun.net.httpserver.ServerImpl$Exchange.run (ServerImpl.java:657)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1167)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:641)
    at java.lang.Thread.run (Thread.java:844)
[WARNING] 
java.lang.NoSuchFieldError: timestampMs
    at io.prometheus.client.exporter.common.TextFormat.write004 (TextFormat.java:49)
    at io.prometheus.client.exporter.HTTPServer$HTTPMetricHandler.handle (HTTPServer.java:59)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:77)
    at sun.net.httpserver.AuthFilter.doFilter (AuthFilter.java:82)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:80)
    at sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle (ServerImpl.java:685)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:77)
    at sun.net.httpserver.ServerImpl$Exchange.run (ServerImpl.java:657)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1167)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:641)
    at java.lang.Thread.run (Thread.java:844)
[WARNING] 
java.lang.NoSuchFieldError: timestampMs
    at io.prometheus.client.exporter.common.TextFormat.write004 (TextFormat.java:49)
    at io.prometheus.client.exporter.HTTPServer$HTTPMetricHandler.handle (HTTPServer.java:59)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:77)
    at sun.net.httpserver.AuthFilter.doFilter (AuthFilter.java:82)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:80)
    at sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle (ServerImpl.java:685)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:77)
    at sun.net.httpserver.ServerImpl$Exchange.run (ServerImpl.java:657)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1167)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:641)
    at java.lang.Thread.run (Thread.java:844)
```

**Comment by songy23** [comment]
This seems to be related to https://github.com/prometheus/client_java/pull/337. Prometheus added the optional `timestamp` field recently this year, and the Prometheus exporter doesn't set this field. 

However, since Prometheus specified `timestamp` as an optional field, I would not expect to see so many warnings when `timestamp` is unset. I'll create an issue to Prometheus to follow up.

**Comment by songy23** [comment]
Filed https://github.com/prometheus/client_java/issues/384 to follow up.

**Comment by songy23** [comment]
Hi @odeke-em , the Prometheus team thought recompiling the project should resolve this:

> That warning is not coming from us, it's from Java. From a quick search it sounds like your users need to recompile their projects from scratch.

Could you try this?

**Comment by seedily** [comment]
Maybe you dependency simpleclient_common-0.3.0.jar, which simple-client:0.2.0 does not support, especially in TextFormat.java:49, it should really throw NoSuchFieldError: timestampMs

**Comment by songy23** [comment]
Hi @odeke-em I just made a `0.14.0-SNAPSHOT` release of OC-Java that upgraded to the latest Prometheus version. I made a PR to update the dependencies in your demo code: https://github.com/orijtech/opencensus-for-grpc-java-developers/pull/1. Could you try running the demo application with the updated dependencies and see how it works?

**Comment by odeke-em** [comment]
>Hi @odeke-em , the Prometheus team thought recompiling the project should resolve this:

Thank you for the followup @songy23! Oh yes, I have done that quite a number of times, recompiling from scratch. Perhaps they mean something else? Does running the repro that I produced work alright on your side? With Prometheus also started too?

 >Maybe you dependency simpleclient_common-0.3.0.jar, which simple-client:0.2.0 does not support, especially in TextFormat.java:49, it should really throw NoSuchFieldError: timestampMs

Thank you @seedily for the recommendation. @songy23 what do you think?

>Hi @odeke-em I just made a 0.14.0-SNAPSHOT release of OC-Java that upgraded to the latest Prometheus version. I made a PR to update the dependencies in your demo code: orijtech/opencensus-for-grpc-java-developers#1. Could you try running the demo application with the updated dependencies and see how it works?

Cool, but unfortunately I just tried again and it still crashes
```shell
java.lang.NoSuchFieldError: timestampMs
    at io.prometheus.client.exporter.common.TextFormat.write004 (TextFormat.java:49)
    at io.prometheus.client.exporter.HTTPServer$HTTPMetricHandler.handle (HTTPServer.java:59)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:77)
    at sun.net.httpserver.AuthFilter.doFilter (AuthFilter.java:82)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:80)
    at sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle (ServerImpl.java:685)
    at com.sun.net.httpserver.Filter$Chain.doFilter (Filter.java:77)
    at sun.net.httpserver.ServerImpl$Exchange.run (ServerImpl.java:657)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1167)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:641)
    at java.lang.Thread.run (Thread.java:844)
```

**Comment by songy23** [comment]
Hmmm I did exactly what @seedily suggested in `0.14.0-SNAPSHOT` release, and now the Prometheus versions in your demo should be consistent after https://github.com/prometheus/client_java/issues/384.

I'll try to reproduce this error later today. If I cannot find anything wrong on OC side, I'll turn to Prometheus for help.

**Comment by songy23** [comment]
Closing - now that we published the v0.14.0 release, please use v0.14.0 OC-Java and 0.4.0 Prometheus simpleclient, the Prometheus versions should be consistent now. Please reopen if updating the versions doesn't resolve this issue.
